src/Algebra/Graph/Algorithm/Scc.hs

```diff
@@ -0,0 +1,64 @@
+module Algebra.Graph.Algorithm.Scc where
+
+import           Algebra.Graph
+import           Algebra.Graph.Algorithm.Internal
+import           Data.Maybe
+import qualified Data.Set                         as Set
+
+-- | O(n^2 * s)
+-- Groups elements into graph of a graph by rule (first function)
+--
+-- >>> let g = (1 + 2) * (3 + 4)
+-- >>> let f x y = even (x + y)
+-- >>> groupBy f g
+-- Overlay (Vertex (Connect (Vertex 1) (Vertex 3))) (Vertex (Connect (Vertex 2) (Vertex 4)))
+groupBy :: Ord a => (a -> a -> Bool) -> Graph a -> Graph (Graph a)
+groupBy f g = foldl1
+  (\x y -> if isAdjacentGraphs x y then connect x y else overlay x y)
```

Comment:
Why `isAdjacentGraphs`? It should be `f`. The idea of `groupBy` is to be more general, and then to use it as `groupBy isAdjacentGraphs` for `scc`.

Reply:
Not really, `f` is used in `groupToList` function. 

- First of all, we apply `groupToList` function on a given graph (`Graph a`) and get list of list of vertices (`[[a]]`) which represents our groups. On this step, we got groups of entities but lost all connection data.
- On the next step, we apply `constructGroup` function to each group (`[a]`) to transform it to a graph (`Graph a`). After this step we have a list of graphs (`[Graph a]`), each graph is one group.
- Also in the previous step, we wrap every `Graph a` by a `Vertex` and got `[Vertex (Graph a)]`.
- And one the final step we check whether groups are connected or not by `isAdjacentGraphs` function. And base on it we either apply `connect` or `overlay` on these vertices.